### PR TITLE
[GOBBLIN-2123] create reevaluate dag action for jobs in state PENDING_RETRY

### PR DESCRIPTION
**Pull request by arjun4084346**
…ProcEngine is enabled.

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-2123

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
- add PENDING_RETRY to FINISHED_STATUSES list, so that a reevaluate dag action is created for it that will submit another execution for this job

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
added tests in ReevaluateDagProcTest

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java [hunk 1]
```diff
@@ -41,7 +41,7 @@
 @Slf4j
 public class FlowStatusGenerator {
   public static final List<String> FINISHED_STATUSES = Lists.newArrayList(ExecutionStatus.FAILED.name(),
-      ExecutionStatus.COMPLETE.name(), ExecutionStatus.CANCELLED.name());
+      ExecutionStatus.COMPLETE.name(), ExecutionStatus.CANCELLED.name(), ExecutionStatus.PENDING_RETRY.name());
```
Comment (phet) [inline gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java:44]: this is used in a number of places. is it acceptable to change the semantics for all of them?  e.g. is it OK here in `MysqlDMStateStore`:
```
  public boolean hasRunningJobs(DagManager.DagId dagId) throws IOException {
    return getDagNodes(dagId).stream()
        .anyMatch(node -> !FlowStatusGenerator.FINISHED_STATUSES.contains(node.getValue().getExecutionStatus().name()));
  }
```
?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/spec/JobExecutionPlanListDeserializer.java [hunk 2]
```diff
@@ -110,6 +110,8 @@ public List<JobExecutionPlan> deserialize(JsonElement json, Type typeOfT, JsonDe
 
       JobExecutionPlan jobExecutionPlan = new JobExecutionPlan(jobSpec, specExecutor);
       jobExecutionPlan.setExecutionStatus(executionStatus);
+      jobExecutionPlan.setCurrentGeneration(serializedJobExecutionPlan.get(SerializationConstants.CURRENT_GENERATION_KEY).getAsInt());
+      jobExecutionPlan.setCurrentAttempts(serializedJobExecutionPlan.get(SerializationConstants.CURRENT_ATTEMPTS_KEY).getAsInt());
```
Comment (phet) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/spec/JobExecutionPlanListDeserializer.java:114]: on a meta level, this PR would be more cohesive as two, one for PENDING_RETRY and the other for (de)serialization

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java (hunk 1 above)
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java:44]: The code is like 
```
if (jobStatus.isShouldRetry()) {
      XXX
    } else if (!dagManagementStateStore.hasRunningJobs(getDagId())) {
      YYY
    }
```
so both code parts do not collide. Also it is kinda ok for `hasRunningJobs` to return `true` when status is PENDING_RETRY, isn't it?

Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java:44]: hmm, actually it should not go to that warn message, because `flowEvent` field is never set to PENDING_RETRY. in fact PENDING_RETRY does not correspond to any "flow event", this is a case when GTE name does not have any execution status

Comment (umustafi) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java:44]: Why use this PENDING_RETRY status at all if Reevaluate is going to use the # of retry attempts to determine if anything to retry for the job? can you summarize how this particular status is used in the PR desc?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/spec/JobExecutionPlanListSerializer.java [hunk 3]
```diff
@@ -82,6 +82,8 @@ public JsonElement serialize(List<JobExecutionPlan> src, Type typeOfSrc, JsonSer
 
       String executionStatus = jobExecutionPlan.getExecutionStatus().name();
       jobExecutionPlanJson.addProperty(SerializationConstants.EXECUTION_STATUS_KEY, executionStatus);
+      jobExecutionPlanJson.addProperty(SerializationConstants.CURRENT_GENERATION_KEY, jobExecutionPlan.getCurrentGeneration());
```
Comment (umustafi) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/spec/JobExecutionPlanListSerializer.java:85]: are these two fields always present in jobExecutionPlan or do they need a default value?

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProcTest.java [hunk 4]
```diff
@@ -263,6 +262,47 @@ public void testMultipleNextJobToRun() throws Exception {
...
+        reEvaluateDagProc = new ReevaluateDagProc(new ReevaluateDagTask(new DagActionStore.DagAction(flowGroup, flowName,
+        flowExecutionId, "job0", DagActionStore.DagActionType.REEVALUATE), null,
+        dagManagementStateStore, mockedDagProcEngineMetrics));
+    reEvaluateDagProc.process(dagManagementStateStore, mockedDagProcEngineMetrics);
+
+    int numOfLaunchedJobs = 1; // only the current job
+    // only the current job should have run
+    Mockito.verify(specProducers.get(0), Mockito.times(1)).addSpec(any());
+
+    specProducers.stream().skip(numOfLaunchedJobs) // separately verified `specProducers.get(0)`
```
Comment (umustafi) [inline gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProcTest.java:295]: is this due to reevaluate dagProc relaunching the job? can u make that more clear with a comment this test is hard to follow bc of the detailed logic being tested here

**Review by phet — APPROVED** [review]

File: gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java (hunk 1 above)
Comment (arjun4084346) [reply gobblin-runtime/src/main/java/org/apache/gobblin/service/monitoring/FlowStatusGenerator.java:44]: Yea, we can actually make it work without PENDING_RETRY. That was the approach in the 1st commit (sorry, it lost because I mistakenly force pushed another squashed commit)
We can completely remove PENDING_RETRY. But in that approach,  job's status will be shown 'FAILED' between both the attempts,, Kip preferred 'PENDING_RETRY'...

File: gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProcTest.java (hunk 4 above)
Comment (arjun4084346) [reply gobblin-service/src/test/java/org/apache/gobblin/service/modules/orchestration/proc/ReevaluateDagProcTest.java:295]: yes, ok, updated the comment
